### PR TITLE
chore: add design system linter

### PR DESCRIPTION
**Pull request by elevatebart**
Closes DESIGN-1567(clickup)

With the components from the design system being progressively available, we want to start replacing `@frontend-shared` primitives with their design system counterparts.
To softly inform devs that they are using a deprecated components, we can use an ESlint rule to inform when a "deprecated" component is used.

For example in `packages/app/src/components/SpecPatternModal.vue`

```vue
<script lang="ts" setup>
// This line will return a warning propting you to replace it with the design systems
import Button from '../../../frontend-shared/src/components/Button.vue'
</script>
<template>
   <Button> {{ t('createSpec.updateSpecPattern') }}</Button>
</template>
```

Since eslint resolves modules like TypeScript, it will also warn on the following code.

```ts
import Button from '@cy/components/Button.vue'
```

### CODE REVIEW

**Comment by elevatebart** [comment]
@mike-plummer I believe this is related to your Design System Buttons spike.

**Review by lmiller1990 — APPROVED** [review]
Looks good, needs rebase! 🙏

File: packages/reporter/src/attempts/attempt-model.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ export default class Attempt {
   @observable _state?: TestState | null = null
   @observable _invocationCount: number = 0
   @observable invocationDetails?: FileDetails
-  @observable hookCount: { [name in HookName]: number } = {
+  @observable hookCount: Record<HookName, number> = {
```
Comment (jordanpowell88) [inline packages/reporter/src/attempts/attempt-model.ts:28]: why is this change needed?

**Review by jordanpowell88 — COMMENTED** [review]
So is the plan to remove these components we are applying the lint rules for from `frontend-shared`? 

I ask as I'm currently working on [reducing our Percy snapshots[(https://github.com/cypress-io/cypress-internal/issues/587). This would help derive some of that work.

File: packages/reporter/src/attempts/attempt-model.ts (hunk 1 above)
Comment (elevatebart) [reply packages/reporter/src/attempts/attempt-model.ts:28]: The linting was complaining about `name` not being used.
Prior to this PR no linting was done at build time.

**Comment by elevatebart** [comment]
@jordanpowell88 yes, the plan is to re-use as many components from the shared Design System as possible.
So if a new component is available and can replace one of yours, I will add it to the list with another PR. 
This way when you come across a component that can be refactored, the documentation is right there.

**Comment by lmiller1990** [comment]
I think we need to wait for the components to have DS parity before merging this @elevatebart, does that sound right? Eg - you cannot fully use `<Button>` from the DS yet, since it has not got fully parity, see https://github.com/cypress-io/cypress/pull/26760 for more on that.

**Comment by elevatebart** [comment]
I am cool with putting it back in draft if we have a date for figuring out the button issues. 

Since the PR is meant to help you deal with the migration, I would hate for it to have the opposite effect. 

I already have solutions and recommendations about the issues found by @mike-plummer on the other PR.

I want to help! I hope to hear from y'all soon.

**Comment by lmiller1990** [comment]
One way to do this might be adding the lint rules incrementally alongside the migration PR. We could add the button one with the button PR, etc.

This sprint is fairly busy, but will revisit in the final sprint of June.

**Comment by warrensplayer** [comment]
@lmiller1990 I agree that it would make sense to address each component with its own PR.  I think we need a chance to fully test out the functionality of the design system component with what the app components do today and decide if we want to:
* update the design system to add any missing functionality 
* keep a "wrapping" component in the app to wrap the design system component and add specific app functionality in the wrapper

At that point, we can decide if linting rules are needed.

**Comment by elevatebart** [comment]
Hi @warrensplayer, Hi @lmiller1990,

Here are my 2 cents: This PR does not add any blocking linting rules. 
It only brings up warnings for known issues.
I would prefer that this PR gets merged as is then change each rule to be an error when the proper PR gets merged.

My reasoning: Using the components in Frontend Shared is the "easiest" for new contributors internal as external. If they want to create a new Button, they will use the copy/paste method. This will spread the usage of these tools and make the "non-consistency problem" worse.

I want to understand more what friction/problem this PR would generate.

**Comment by lmiller1990** [comment]
My opinion is to *update the design system to add any missing functionality* where possible, especially for simple use cases. If we just keep wrapping things, we end up with this weird meta design system - a design system built on a design system. 

It also feels wrong to deprecate a component via a linting rule if no known replacement exists. Most developers are trained to avoid deprecated things, so I'd rather add these rules 1 by 1 as the design system comes up to parity. I've filed a few issues there, mostly easy ones, so this shouldn't be a huge lift.

**Comment by elevatebart** [comment]
@lmiller1990 It's still unclear what problem it poses. 
My objective with this PR is to incentivize using the design system instead of copy pasting what is already there.
It is also when one does work to update a component to also know that they are working in deprecated code.

I know it puts pressure on the dev to see the yellow squiggles. Is that the pressure/problem mental tension you are afraid of?

**Comment by marktnoonan** [comment]
I'm in favor of this in general. But I think when a lint rule (or warning) is in place, it must be possible to follow it, which is not the case if the components are not at parity. I have a bunch of other thoughts on how we might adopt the design system, I will try to come back to this today.

**Comment by lmiller1990** [comment]
> I know it puts pressure on the dev to see the yellow squiggles

Yeah, I don't like the idea of using something my linter is telling me not to. I think it's not that much work to get some of the components fully migrated, and add these rules as components are migrated. I wonder what the quickest full migration component would be, maybe button or checkbox.

**Comment by elevatebart** [comment]
@lmiller1990 Once a component is fully migrated, I expect the outdated component to be deleted. 
Then the lint rule will point to a component that does not exist and typescript will do the work for us.

If we decide to add the lint rules after migration, I would suggest we abandon the lint rules entirely.

I am sorry about being dramatic.

**Comment by marktnoonan** [comment]
> If we decide to add the lint rules after migration, I would suggest we abandon the lint rules entirely.

Agreed, the lint warning only makes sense imo if we are in this state:

1. DS component is ready to be fully used in the app - meaning a developer _can_ follow the warning and use the DS
1. App has a legacy component in frontend-shared with the same purpose, that is still in use

But the main way validate number 1, is to actually do the work and find the edge cases, then fix them in the DS.

I'm going to throw my suggestions for how we organize this into an internal thread for feedback.

**Review by ryanjwilke — APPROVED** [review]
LGTM 👍

**Comment by elevatebart** [comment]
To avoid confusion, @marktnoonan and I decided to adjust the linter a bit:
- We split the configuration for deprecated components in a separate `.eslintrc.design-system.js` file
- The linting process will not show squiggles to developers in vscode
- The linting is going to be run for comparison in CI to measure progress over time
- If devs want to see their progress or work on design system migration they will have a yarn command to run the lint

**Comment by marktnoonan** [comment]
When that change is in ☝️ I'll be happy to approve this PR. I can see the usefulness in being able to check on this from time to time to understand the trend in usage of components that could be replaced by the DS and measure progress towards adoption. And if it's not giving me a confusing warning in VSCode, I don't see a downside.

**Review by marktnoonan — APPROVED** [review]

**Comment by marktnoonan** [comment]
@elevatebart looks like it is possible for "Linting design system migration" step to error and fail the linux-lint step in CI, is there a way to prevent this?

**Comment by elevatebart** [comment]
I thought I did prevent it. That's the purpose of the `--no-eslintrc` argument. I will look at it tomorrow then.
I agree there should not be such issues.

**Comment by lmiller1990** [comment]
We can merge this soon, we have some design system work coming next sprint https://github.com/cypress-io/cypress-internal/issues/614 we can migrate some components, and deprecate the migrated ones using the code in this PR.
